File: Data Structures & Algorithms/reorder-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        ListNode* temp = head;

        ListNode* fast = head->next;
        ListNode* slow = head;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }

        ListNode* secondPiece = slow->next;
        slow->next = nullptr;

        ListNode* next = nullptr;
        ListNode* prev = nullptr;

        while (secondPiece) {
            next = secondPiece->next;
            secondPiece->next = prev;
            prev = secondPiece;
            secondPiece = next;
        }

        ListNode* firstPiece = head;
        secondPiece = prev; 

        while(secondPiece) {
            ListNode* temp1 = firstPiece->next;
            ListNode* temp2 = secondPiece->next;

            firstPiece->next = secondPiece;
            secondPiece->next = temp1;
            firstPiece = temp1;
            secondPiece = temp2;
        }
    }
};

File: Data Structures & Algorithms/palindrome-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        slow = fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        temp = slow
        prev = None
        while temp:
            nxt = temp.next
            temp.next = prev
            prev = temp
            temp = nxt
        while prev:
            if prev.val != head.val:
                return False
            prev = prev.next
            head = head.next
        return True